dm/master/openapi.go

```diff
@@ -133,17 +135,99 @@ func (s *Server) DMAPIDeleteSource(ctx echo.Context, sourceName string) error {
 
 // DMAPIStartRelay url is:(POST /api/v1/sources/{source-id}/relay).
 func (s *Server) DMAPIStartRelay(ctx echo.Context, sourceName string) error {
-	return nil
+	var req openapi.StartRelayRequest
+	if err := ctx.Bind(&req); err != nil {
+		return err
+	}
+	sourceCfg := s.scheduler.GetSourceCfgByID(sourceName)
+	if sourceCfg == nil {
+		return terror.ErrSchedulerSourceCfgNotExist.Generate(sourceName)
+	}
+	needUpdate := false
+	// update relay related in source cfg
+	if req.RelayBinlogName != nil && sourceCfg.RelayBinLogName != *req.RelayBinlogName {
+		sourceCfg.RelayBinLogName = *req.RelayBinlogName
+		needUpdate = true
+	}
+	if req.RelayBinlogGtid != nil && sourceCfg.RelayBinlogGTID != *req.RelayBinlogGtid {
+		sourceCfg.RelayBinlogGTID = *req.RelayBinlogGtid
+		needUpdate = true
+	}
+	if req.RelayDir != nil && sourceCfg.RelayDir != *req.RelayDir {
+		sourceCfg.RelayDir = *req.RelayDir
+		needUpdate = true
+	}
+	if purge := req.Purge; purge != nil {
+		if purge.Expires != nil && sourceCfg.Purge.Expires != *purge.Expires {
+			sourceCfg.Purge.Expires = *purge.Expires
+			needUpdate = true
+		}
+		if purge.Interval != nil && sourceCfg.Purge.Interval != *purge.Interval {
+			sourceCfg.Purge.Interval = *purge.Interval
+			needUpdate = true
+		}
+		if purge.RemainSpace != nil && sourceCfg.Purge.RemainSpace != *purge.RemainSpace {
+			sourceCfg.Purge.RemainSpace = *purge.RemainSpace
+			needUpdate = true
+		}
+	}
+	if needUpdate {
+		// update current source relay config before start relay
+		if err := s.scheduler.UpdateSourceCfg(sourceCfg); err != nil {
+			return err
+		}
+	}
+	return s.scheduler.StartRelay(sourceName, []string{req.WorkerName})
 }
 
 // DMAPIStopRelay url is:(DELETE /api/v1/sources/{source-id}/relay).
 func (s *Server) DMAPIStopRelay(ctx echo.Context, sourceName string) error {
-	return nil
+	var req openapi.WorkerNameRequest
+	if err := ctx.Bind(&req); err != nil {
+		return err
+	}
+	return s.scheduler.StopRelay(sourceName, []string{req.WorkerName})
 }
 
 // DMAPIGetSourceStatus url is:(GET /api/v1/sources/{source-id}/status).
 func (s *Server) DMAPIGetSourceStatus(ctx echo.Context, sourceName string) error {
-	return nil
+	sourceCfg := s.scheduler.GetSourceCfgByID(sourceName)
+	if sourceCfg == nil {
+		return terror.ErrSchedulerSourceCfgNotExist.Generate(sourceName)
+	}
+	resp := openapi.SourceStatus{SourceName: sourceCfg.SourceID}
+	worker := s.scheduler.GetWorkerBySource(sourceName)
+	// current this source not bound to any worker
+	if worker == nil {
+		return ctx.JSON(http.StatusOK, resp)
+	}
+	resp.WorkerName = worker.BaseInfo().Name
+	// get status from worker
+	workerReq := &workerrpc.Request{
+		Type:        workerrpc.CmdQueryStatus,
+		QueryStatus: &pb.QueryStatusRequest{Name: ""},
+	}
+	workerResp, err := worker.SendRequest(ctx.Request().Context(), workerReq, s.cfg.RPCTimeout)
+	if err != nil {
+		return terror.ErrOpenAPICommonError.New(err.Error())
```

Comment:
maybe this will lose the walkarond field

Reply:
yes it will, but i think's this will